File: backend/app/models/enums.py
"""Enumerations used across the domain models.

The ``value`` of each member is the canonical machine value stored in the
database. Human readable Russian labels are provided through :meth:`label` and
are primarily used for exports (Excel / PDF) where the backend renders text.
The Vue frontend additionally localises these values through Vue I18n.
"""
from __future__ import annotations

import enum


class UserRole(str, enum.Enum):
    """System roles controlling access to modules."""

    ADMINISTRATOR = "administrator"
    RENTAL_MANAGER = "rental_manager"
    CASHIER = "cashier"
    OPERATOR = "operator"

    @property
    def label(self) -> str:
        return {
            "administrator": "Администратор",
            "rental_manager": "Менеджер аренды",
            "cashier": "Кассир",
            "operator": "Оператор",
        }[self.value]


class ClientStatus(str, enum.Enum):
    ACTIVE = "active"
    PENDING_VERIFICATION = "pending_verification"
    BLACKLISTED = "blacklisted"

    @property
    def label(self) -> str:
        return {
            "active": "Активен",
            "pending_verification": "Ожидает проверки",
            "blacklisted": "В чёрном списке",
        }[self.value]


class CarStatus(str, enum.Enum):
    AVAILABLE = "available"
    RESERVED = "reserved"
    RENTED = "rented"
    MAINTENANCE = "maintenance"

    @property
    def label(self) -> str:
        return {
            "available": "Доступен",
            "reserved": "Забронирован",
            "rented": "В аренде",
            "maintenance": "На обслуживании",
        }[self.value]


class ReservationStatus(str, enum.Enum):
    RESERVED = "reserved"
    CONFIRMED = "confirmed"
    CANCELLED = "cancelled"

    @property
    def label(self) -> str:
        return {
            "reserved": "Забронировано",
            "confirmed": "Подтверждено",
            "cancelled": "Отменено",
        }[self.value]


class RentalStatus(str, enum.Enum):
    ACTIVE = "active"
    COMPLETED = "completed"
    CANCELLED = "cancelled"

    @property
    def label(self) -> str:
        return {
            "active": "Активна",
            "completed": "Завершена",
            "cancelled": "Отменена",
        }[self.value]


class PaymentMethod(str, enum.Enum):
    CASH = "cash"
    BANK_TRANSFER = "bank_transfer"
    CARD = "card"

    @property
    def label(self) -> str:
        return {
            "cash": "Наличные",
            "bank_transfer": "Банковский перевод",
            "card": "Карта",
        }[self.value]


class PaymentType(str, enum.Enum):
    DEPOSIT = "deposit"
    RENTAL = "rental"
    PENALTY = "penalty"
    DAMAGE = "damage"
    REFUND = "refund"

    @property
    def label(self) -> str:
        return {
            "deposit": "Депозит",
            "rental": "Оплата аренды",
            "penalty": "Штраф",
            "damage": "Возмещение ущерба",
            "refund": "Возврат средств",
        }[self.value]


class BlacklistReason(str, enum.Enum):
    FRAUD = "fraud"
    VEHICLE_DAMAGE = "vehicle_damage"
    NON_PAYMENT = "non_payment"
    SERIOUS_VIOLATION = "serious_violation"

    @property
    def label(self) -> str:
        return {
            "fraud": "Мошенничество",
            "vehicle_damage": "Повреждение автомобиля",
            "non_payment": "Неоплата",
            "serious_violation": "Серьёзное нарушение",
        }[self.value]

File: backend-go/internal/models/enums.go
package models

// Enumerations used across the domain models. The string value is the canonical
// machine value stored in the database (VARCHAR). Human readable Russian labels
// are exposed through the Label methods and used for exports (Excel / PDF).

// UserRole controls access to modules.
type UserRole string

const (
	RoleAdministrator UserRole = "administrator"
	RoleRentalManager UserRole = "rental_manager"
	RoleCashier       UserRole = "cashier"
	RoleOperator      UserRole = "operator"
)

var userRoleLabels = map[UserRole]string{
	RoleAdministrator: "Администратор",
	RoleRentalManager: "Менеджер аренды",
	RoleCashier:       "Кассир",
	RoleOperator:      "Оператор",
}

// Label returns the Russian display label for the role.
func (r UserRole) Label() string { return userRoleLabels[r] }

// Valid reports whether the role is a known value.
func (r UserRole) Valid() bool { _, ok := userRoleLabels[r]; return ok }

// AllUserRoles returns every role in declaration order.
func AllUserRoles() []UserRole {
	return []UserRole{RoleAdministrator, RoleRentalManager, RoleCashier, RoleOperator}
}

// ClientStatus represents the verification state of a client.
type ClientStatus string

const (
	ClientActive              ClientStatus = "active"
	ClientPendingVerification ClientStatus = "pending_verification"
	ClientBlacklisted         ClientStatus = "blacklisted"
)

var clientStatusLabels = map[ClientStatus]string{
	ClientActive:              "Активен",
	ClientPendingVerification: "Ожидает проверки",
	ClientBlacklisted:         "В чёрном списке",
}

func (s ClientStatus) Label() string { return clientStatusLabels[s] }
func (s ClientStatus) Valid() bool   { _, ok := clientStatusLabels[s]; return ok }

// CarStatus represents the availability of a vehicle.
type CarStatus string

const (
	CarAvailable   CarStatus = "available"
	CarReserved    CarStatus = "reserved"
	CarRented      CarStatus = "rented"
	CarMaintenance CarStatus = "maintenance"
)

var carStatusLabels = map[CarStatus]string{
	CarAvailable:   "Доступен",
	CarReserved:    "Забронирован",
	CarRented:      "В аренде",
	CarMaintenance: "На обслуживании",
}

func (s CarStatus) Label() string { return carStatusLabels[s] }
func (s CarStatus) Valid() bool   { _, ok := carStatusLabels[s]; return ok }

func AllCarStatuses() []CarStatus {
	return []CarStatus{CarAvailable, CarReserved, CarRented, CarMaintenance}
}

// ReservationStatus represents the state of a reservation.
type ReservationStatus string

const (
	ReservationReserved  ReservationStatus = "reserved"
	ReservationConfirmed ReservationStatus = "confirmed"
	ReservationCancelled ReservationStatus = "cancelled"
)

var reservationStatusLabels = map[ReservationStatus]string{
	ReservationReserved:  "Забронировано",
	ReservationConfirmed: "Подтверждено",
	ReservationCancelled: "Отменено",
}

func (s ReservationStatus) Label() string { return reservationStatusLabels[s] }
func (s ReservationStatus) Valid() bool   { _, ok := reservationStatusLabels[s]; return ok }

// RentalStatus represents the state of a rental contract.
type RentalStatus string

const (
	RentalActive    RentalStatus = "active"
	RentalCompleted RentalStatus = "completed"
	RentalCancelled RentalStatus = "cancelled"
)

var rentalStatusLabels = map[RentalStatus]string{
	RentalActive:    "Активна",
	RentalCompleted: "Завершена",
	RentalCancelled: "Отменена",
}

func (s RentalStatus) Label() string { return rentalStatusLabels[s] }
func (s RentalStatus) Valid() bool   { _, ok := rentalStatusLabels[s]; return ok }

func AllRentalStatuses() []RentalStatus {
	return []RentalStatus{RentalActive, RentalCompleted, RentalCancelled}
}

// PaymentMethod represents how a payment was made.
type PaymentMethod string

const (
	PaymentCash         PaymentMethod = "cash"
	PaymentBankTransfer PaymentMethod = "bank_transfer"
	PaymentCard         PaymentMethod = "card"
)

var paymentMethodLabels = map[PaymentMethod]string{
	PaymentCash:         "Наличные",
	PaymentBankTransfer: "Банковский перевод",
	PaymentCard:         "Карта",
}

func (m PaymentMethod) Label() string { return paymentMethodLabels[m] }
func (m PaymentMethod) Valid() bool   { _, ok := paymentMethodLabels[m]; return ok }

// PaymentType represents the purpose of a payment.
type PaymentType string

const (
	PaymentTypeDeposit PaymentType = "deposit"
	PaymentTypeRental  PaymentType = "rental"
	PaymentTypePenalty PaymentType = "penalty"
	PaymentTypeDamage  PaymentType = "damage"
	PaymentTypeRefund  PaymentType = "refund"
)

var paymentTypeLabels = map[PaymentType]string{
	PaymentTypeDeposit: "Депозит",
	PaymentTypeRental:  "Оплата аренды",
	PaymentTypePenalty: "Штраф",
	PaymentTypeDamage:  "Возмещение ущерба",
	PaymentTypeRefund:  "Возврат средств",
}

func (t PaymentType) Label() string { return paymentTypeLabels[t] }
func (t PaymentType) Valid() bool   { _, ok := paymentTypeLabels[t]; return ok }

// BlacklistReason represents why a client was blacklisted.
type BlacklistReason string

const (
	ReasonFraud            BlacklistReason = "fraud"
	ReasonVehicleDamage    BlacklistReason = "vehicle_damage"
	ReasonNonPayment       BlacklistReason = "non_payment"
	ReasonSeriousViolation BlacklistReason = "serious_violation"
)

var blacklistReasonLabels = map[BlacklistReason]string{
	ReasonFraud:            "Мошенничество",
	ReasonVehicleDamage:    "Повреждение автомобиля",
	ReasonNonPayment:       "Неоплата",
	ReasonSeriousViolation: "Серьёзное нарушение",
}

func (r BlacklistReason) Label() string { return blacklistReasonLabels[r] }
func (r BlacklistReason) Valid() bool   { _, ok := blacklistReasonLabels[r]; return ok }

func AllBlacklistReasons() []BlacklistReason {
	return []BlacklistReason{ReasonFraud, ReasonVehicleDamage, ReasonNonPayment, ReasonSeriousViolation}
}
